pkg/ccl/backupccl/restore_processor_planning.go

```diff
@@ -264,8 +266,23 @@ func distRestore(
 	g.GoCtx(func(ctx context.Context) error {
 		metaFn := func(_ context.Context, meta *execinfrapb.ProducerMetadata) error {
 			if meta.BulkProcessorProgress != nil {
-				// Send the progress up a level to be written to the manifest.
-				progCh <- meta.BulkProcessorProgress
+				if meta.BulkProcessorProgress.Drained {
+					testingKnobs := execCtx.ExecCfg().BackupRestoreTestingKnobs
+					if testingKnobs != nil && testingKnobs.RunAfterRestoreProcDrains != nil {
+						testingKnobs.RunAfterRestoreProcDrains()
+					}
+					// procCompleteCh will not have any listeners if the restore job
+					// needs to replan due to lagging nodes. In that case, do not
+					// send to the channel.
```

Comment:
Might the same be true of progCh below? I think ctxgroup.GoAndWait will cancel the context once one of the goroutines returns which could cause the reader of progCh to exit early as well.

Perhaps both of these should be selects that also have a <-ctx.Done() case.

Reply:
Added `<-ctx.Done()` for both cases!